### ISSUE TITLE
localhost is not a good default "publicHost" for development

### ISSUE THREAD

**Issue by dkbast** [serverpod/serverpod#2127]
**Describe the bug**
The default config uses localhost as the server address - this has multiple issues when working with a physical device and not just with an emulator:

1) A physical phone cannot connect to the 'localhost' of the host machine. We can get around this by passing the local ip address to the phone at runtime and use to initialise the client.

My current workaround for 1) is this:

```bash
flutter run --pid-file=${FLUTTER_PID_FILE} --dart-define=SERVER=http://$(shell ipconfig getifaddr en0):8080/
``` 

2) Even if passing the ip to configure the client correctly, the server will still think it is serving as "localhost" which makes it impossible to get assets from the server:

```
CachedNetworkImageProvider("http://localhost:8080/serverpod_cloud_storage?method=file&path=serverpod%2Fuser
_images%2F1-1.jpg",
``` 

**To Reproduce**
1. Create a new serverpod project
2. Attach physical device
3. Pass the hosts ip address to the client in the app
4. Add UserImageButton and upload an image 

**Expected behavior**
There should be a developer friendly way to run a server and develop on a physical device. Since the 'publicHost' is declared in the config file its not as easily fixed as the first problem, but it would be great if there would be a serverpod cli command for both, running a serverpod flutter app with the correct publicHost as well as starting a serverpod server with a correct publicHost. Correct == discoverable in the network outside of the machine. 

**Serverpod versions**
 -  _CLI Version:_ 1.2.6
 - _Library version:_ 1.2.6

**Platform information**
MacOS ARM & iOS

**Comment by lukehutch** [comment]
I had both issues too: #979

**Comment by dkbast** [comment]
Yes looks like this has not been fixed properly :/ - I think that the yaml based configuration is great for stuff that is the same between instances but that dev/staging/prod might be solved in a better way - maybe by having three different entry points, similar to the very_good create template for flutter apps

**Comment by Code-Remote** [comment]
I just set up a serverpod and implemented authentication. I'm running on an Android emulator but my login image can't be served to me. In the docs is says to also change the `publicHost` of your apiServer in the `config/development.yaml`

In my flutter code I've changed the client to 10.0.2.2 and I've changed it to 10.0.2.2 in the config file but on my Android emulator  my image still does not get served.

```
apiServer:
  port: 8080
  publicHost: 10.0.2.2 #changed from local host
  publicPort: 8080
  publicScheme: http
  ```

In the logcat I still see that `localhost` is being used? Is this a related issue or am I doing something wrong in the setup? I Do agree that if I have to test my app on multiple devices (emulator, simulator and physical device) it can be quite a burden.
```
The following _ClientSocketException was thrown resolving an image codec:
ClientException with SocketException: Connection refused (OS Error: Connection refused, errno = 111), address = localhost, port = 48328, uri=http://localhost:8080/serverpod_cloud_storage?method=file&path=serverpod%2Fuser_images%2F1-1.jpg

When the exception was thrown, this was the stack: 
Image provider: CachedNetworkImageProvider("http://localhost:8080/serverpod_cloud_storage?method=file&path=serverpod%2Fuser_images%2F1-1.jpg", scale: 1.0)
Image key: CachedNetworkImageProvider("http://localhost:8080/serverpod_cloud_storage?method=file&path=serverpod%2Fuser_images%2F1-1.jpg",
```

**Comment by SandPod** [comment]
I believe this could be solved by having the possibility to reflect the host used to connect to the server in the request when building the public url.

**Comment by Code-Remote** [comment]
> I believe this could be solved by having the possibility to reflect the host used to connect to the server in the request when building the public url.

I'm not really sure if I understand your answer, does that mean that the image URL was generated before i changed my publicHost address?  Should my request URL by default be dynamically build and get it from the config file to see what the current webserver address is?

Do you maybe have a practical example of what to do when i switch from prod / dev while using an android emulator vs a real device?
